File: C++/1371.cpp
#include <string>
#include <unordered_map>

class Solution {
public:
    int findTheLongestSubstring(std::string s) {
        std::string vowels = "aeiou";
        int ans = 0;
        int mask = 0;
        std::unordered_map<int, int> mask_to_idx;
        mask_to_idx[0] = -1;

        for(int i = 0; i < s.size(); ++i){
            char c = s[i];
            if(vowels.find(c) != std::string::npos){
                mask ^= (c - 'a') + 1;
            }
            if (mask_to_idx.find(mask) != mask_to_idx.end()) {
                ans = std::max(ans, i - mask_to_idx[mask]);
            } else {
                mask_to_idx[mask] = i;
            }
        }
        return ans;
    }
};

// Topic: Array, Bit Manipulation
// Time Complexity: O(n)
// Space Complexity: O(1)

File: Python/1371.py
class Solution:
    def findTheLongestSubstring(self, s: str) -> int:
        # Count, First index, Last idnex
        vowels = 'aeiou'
        ans = 0
        mask = 0
        mask_to_idx ={0: -1}

        for idx, c in enumerate(s):
            if c in vowels:
                mask ^= ord(c)-ord('a') + 1
            if mask in mask_to_idx:
                ans = max(ans, idx-mask_to_idx[mask])
            else:
                mask_to_idx[mask] = idx
        
        return ans

# Topic: Bit Manipulation
# Time complexity: O(n)
# Space complexity: O(1)